packages/caravan-fees/src/cpfp.ts

```diff
@@ -0,0 +1,317 @@
+      address: changeAddress,
+      amountSats: "0", // Initial amount, will be adjusted later
+    }),
+  );
+
+  /**
+   * Step 6: Add additional inputs to cover the CPFP fee requirements
+   *
+   * This step implements a simplified version of the "Child-pays-for-parent" (CPFP)
+   * transaction selection logic, inspired by Bitcoin Core's implementation
+   * (https://github.com/bitcoin/bitcoin/pull/7600).
+   *
+   * In Bitcoin Core, transactions are sorted in the mempool using a modified
+   * feerate that considers ancestors. During block creation, transactions are
+   * considered in this order. As transactions are selected, a separate map tracks
+   * the new feerate-with-ancestors for in-mempool descendants.
+   *
+   * Our simplified approach:
+   * 1. We start with the parent transaction's fee and vsize.
+   * 2. We iteratively add inputs to the child transaction.
+   * 3. After each input addition, we recalculate the combined fee rate of the
+   *    parent and child transactions.
+   * 4. We continue adding inputs until the combined fee rate meets or exceeds
+   *    the target fee rate.
+   *
+   * This method ensures that the child transaction provides enough fee to incentivize
+   * miners to include both the parent and child transactions in a block, effectively
+   * "bumping" the fee of the parent transaction.
+   *
+   * Note: While this approach is simpler than Bitcoin Core's full implementation,
+   * it achieves the core goal of CPFP: ensuring the combined package (parent + child)
+   * is attractive for miners to include in a block.
+   */
+
+  for (const utxo of availableInputs) {
+    if (isCPFPFeeSatisfied(txAnalyzer, childTxTemplate)) {
+      break; // Stop adding inputs once CPFP fee requirements are met
+    }
+    // Skip if this UTXO is already added
+    if (
+      childTxTemplate.inputs.some(
+        (input) => input.txid === utxo.txid && input.vout === utxo.vout,
+      )
+    ) {
+      continue;
+    }
+    childTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(utxo));
+  }
+
+  // Step 7: Calculate and set the change output amount
+  const totalInputAmount = new BigNumber(childTxTemplate.totalInputAmount);
+  const fee = new BigNumber(childTxTemplate.targetFeesToPay);
+  const changeAmount = totalInputAmount.minus(fee);
+
+  // Check if change amount is above dust threshold
+  if (changeAmount.gt(dustThreshold)) {
+    // Change amount is valid, set it as the output amount
+    childTxTemplate.adjustChangeOutput();
+  } else if (strict) {
+    // Change amount is at or below dust threshold
+
+    // In strict mode, throw an error
+    throw new Error(
+      "Change amount is below the dust threshold. Increase inputs or reduce fee rate.",
+    );
+  } else {
+    // In non-strict mode, warn and proceed with the dust amount
+    console.warn(
+      "Change amount is below dust threshold. Transaction may be less efficient.",
+    );
+    // Set change amount
+    childTxTemplate.adjustChangeOutput();
+  }
```

Comment:
> we can but then we won't be able to throw the error-specific to the strict mode and that warning

Honestly, I think it's worth removing especially since adjustChangeOutput and the validate function below already do good validation and return useful information. 

So I'm thinking at this point, you've added the inputs to cover fees for the CPFP package. If you add the dust check above then you'll stop once you have enough leftover to have one change output with an amount leftover that's not dust. So here what you could do is something like:

```suggestion
  // needsChange returns true if there is enough left over greater than dust. 
  // Since we already added a zero amount output for change, it takes into account the full size of the tx. 
  if (!childTxTemplate.needsChange() && strict) {
        throw new Error("Not enough inputs to create a non-dusty change output in the child transaction")
   }
   
   childTxTemplate.adjustChangeOutput();
```

I don't think the warn is really necessary to be honest anyway (I'm not even sure the later validations would let things pass even if you did). So just throw if strict and let other areas handle it afterwards.

Reply:
thanks for the suggestions ... this really cleans up a lot of code and also handles for the dust Outputs , really nice refactor ... :)
[review-changes: CPFP](https://github.com/caravan-bitcoin/caravan/pull/114/commits/40576591fc50b736bdf91485b2e02eedbd2feb97)